File: public/java/064-partition-equal-subset-sum.java
/*
 * LeetCode Problem #416: Partition Equal Subset Sum
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/partition-equal-subset-sum/
 */
import java.util.*;

class Solution {

    // APPROACH 1: 2D Boolean DP | O(n*sum) time | O(n*sum) space
    // EXPLAIN: dp[i][s] = can we form sum s using first i numbers; classic 0/1 knapsack.
    // WHEN: Use when you need the full table to reconstruct which elements were selected.

    public boolean canPartition_2d(int[] nums) {
        int total = 0;
        for (int n : nums) total += n;
        if (total % 2 != 0) return false;
        int target = total / 2;
        int n = nums.length;
        boolean[][] dp = new boolean[n + 1][target + 1];
        dp[0][0] = true;
        for (int i = 1; i <= n; i++) {
            for (int s = 0; s <= target; s++) {
                dp[i][s] = dp[i - 1][s];
                if (s >= nums[i - 1]) dp[i][s] |= dp[i - 1][s - nums[i - 1]];
            }
        }
        return dp[n][target];
    }

    // APPROACH 2: Space-Optimized 1D DP | O(n*sum) time | O(sum) space
    // EXPLAIN: Reduce to a 1D bitset by iterating the sum dimension in reverse to avoid using same item twice.
    // WHEN: Use as the optimal solution — same time complexity but linear space.

    public boolean canPartition(int[] nums) {
        int total = 0;
        for (int n : nums) total += n;
        if (total % 2 != 0) return false;
        int target = total / 2;
        boolean[] dp = new boolean[target + 1];
        dp[0] = true;
        for (int num : nums) {
            // Traverse backwards so each item is used at most once
            for (int s = target; s >= num; s--) {
                dp[s] |= dp[s - num];
            }
        }
        return dp[target];
    }

    // APPROACH 3: Memoization (Top-Down) | O(n*sum) time | O(n*sum) space
    // EXPLAIN: Recursively include or exclude each item; cache (index, remaining) pairs.
    // WHEN: Top-down style is preferred; memoization avoids redundant sub-problem evaluation.

    public boolean canPartition_memo(int[] nums) {
        int total = 0;
        for (int n : nums) total += n;
        if (total % 2 != 0) return false;
        int target = total / 2;
        Boolean[][] memo = new Boolean[nums.length + 1][target + 1];
        return memoHelper(nums, 0, target, memo);
    }

    private boolean memoHelper(int[] nums, int idx, int rem, Boolean[][] memo) {
        if (rem == 0) return true;
        if (idx >= nums.length || rem < 0) return false;
        if (memo[idx][rem] != null) return memo[idx][rem];
        memo[idx][rem] = memoHelper(nums, idx + 1, rem - nums[idx], memo)
                       || memoHelper(nums, idx + 1, rem, memo);
        return memo[idx][rem];
    }

    // APPROACH 4: Bitset Optimisation | O(n*target/64) time | O(target) space
    // EXPLAIN: Represent reachable sums as a BitSet; shift bits by each num to extend sums.
    // WHEN: Java BitSet provides hardware-level parallelism — fastest practical approach.

    public boolean canPartition_bitset(int[] nums) {
        int total = 0;
        for (int n : nums) total += n;
        if (total % 2 != 0) return false;
        int target = total / 2;
        BitSet bits = new BitSet(target + 1);
        bits.set(0);
        for (int num : nums) {
            BitSet shifted = bits.get(0, target + 1 - num);
            // shift left by num using manual iteration
            for (int i = target; i >= num; i--) {
                if (bits.get(i - num)) bits.set(i);
            }
            if (bits.get(target)) return true;
        }
        return bits.get(target);
    }
}

// Made with Bob

File: public/python/064-partition-equal-subset-sum.py
# LeetCode Problem #416: Partition Equal Subset Sum
# Difficulty: Medium
# Link: https://leetcode.com/problems/partition-equal-subset-sum/

from functools import lru_cache
from typing import List

# ─────────────────────────────────────────────
# APPROACH 1: Recursive Memoization | O(n·sum) time | O(n·sum) space
# EXPLAIN: For each item decide include or exclude; cache (index, remaining target).
# WHEN: Clear top-down derivation of the 0/1 knapsack recurrence.

def canPartition_memo(nums: List[int]) -> bool:
    total = sum(nums)
    if total % 2:
        return False
    target = total // 2

    @lru_cache(maxsize=None)
    def dp(i, rem):
        if rem == 0:
            return True
        if i == len(nums) or rem < 0:
            return False
        return dp(i + 1, rem - nums[i]) or dp(i + 1, rem)

    return dp(0, target)


# APPROACH 2: Bottom-Up DP (1-D bitset) | O(n·sum) time | O(sum) space
# EXPLAIN: Use a boolean set of reachable sums; for each number extend all current sums.
# WHEN: Best practical choice — concise, no recursion, O(sum) space via set / bitmask.

def canPartition_dp(nums: List[int]) -> bool:
    total = sum(nums)
    if total % 2:
        return False
    target = total // 2
    dp = {0}
    for n in nums:
        dp = {s + n for s in dp} | dp
        if target in dp:
            return True
    return False


# APPROACH 3: 2D Bottom-Up DP | O(n·sum) time | O(n·sum) space
# EXPLAIN: dp[i][s] = can first i items form sum s; classic 0/1 knapsack table.
# WHEN: When you need to inspect intermediate dp values or reconstruct the partition.

def canPartition_2d(nums: List[int]) -> bool:
    total = sum(nums)
    if total % 2:
        return False
    target = total // 2
    n = len(nums)
    dp = [[False] * (target + 1) for _ in range(n + 1)]
    for i in range(n + 1):
        dp[i][0] = True
    for i in range(1, n + 1):
        for s in range(1, target + 1):
            dp[i][s] = dp[i - 1][s]
            if s >= nums[i - 1]:
                dp[i][s] = dp[i][s] or dp[i - 1][s - nums[i - 1]]
    return dp[n][target]


# APPROACH 4: 1-D Array DP (reverse iteration) | O(n·sum) time | O(sum) space
# EXPLAIN: Standard 0/1 knapsack with reverse-order inner loop to avoid item reuse.
# WHEN: Most memory-efficient array approach; canonical 0/1 knapsack reduction.

def canPartition_1d(nums: List[int]) -> bool:
    total = sum(nums)
    if total % 2:
        return False
    target = total // 2
    dp = [False] * (target + 1)
    dp[0] = True
    for num in nums:
        for s in range(target, num - 1, -1):
            dp[s] = dp[s] or dp[s - num]
    return dp[target]


# ── quick tests ──────────────────────────────────────────────────────────────
if __name__ == '__main__':
    cases = [
        ([1, 5, 11, 5], True),
        ([1, 2, 3, 5], False),
        ([1, 1], True),
        ([1], False),
        ([3, 3, 3, 4, 5], True),
    ]
    for fn in (canPartition_memo, canPartition_dp, canPartition_2d, canPartition_1d):
        for nums, expected in cases:
            result = fn(nums)
            assert result == expected, (
                f'{fn.__name__}({nums}) = {result}, expected {expected}'
            )
    print('All tests passed.')

# Made with Bob
